docs/3.x/extend/graphql.md

```diff
@@ -0,0 +1,1169 @@
+{
+  widgets {
+    dateCreated @formatDateTime (format: "Y-m-d")
+  }
+}
+```
+
+Craft’s included directives apply exclusively to requested fields, though they may be applied in mutations and numerous parts of the type system.
+
+A directive needs to provide a name, description, and the relevant query location(s) it can be applied. It can optionally specify [arguments](#arguments).
+
+### Example Directive Class
+
+This simple example returns the field value, replacing any instances of `foo` with `bar`.
+
+```php
+namespace mynamespace\gql\directives;
+
+use craft\gql\base\Directive;
+use craft\gql\GqlEntityRegistry;
+use GraphQL\Type\Definition\Directive as GqlDirective;
+
+class BarTheFoo extends Directive
+{
+
+    public static function create(): GqlDirective
+    {
+        if ($type = GqlEntityRegistry::getEntity(self::name())) {
+            return $type;
+        }
+
+        $type = GqlEntityRegistry::createEntity(static::name(), new self([
+            'name' => static::name(),
+            'locations' => [
+                DirectiveLocation::FIELD,
+            ],
+            'description' => 'Replace `foo` with `bar`.',
+            'args' => [],
+        ]));
+
+        return $type;
+    }
+
+    public static function name(): string
+    {
+        return 'fooTheBar';
+    }
+
+    public static function apply($source, $value, array $arguments, ResolveInfo $resolveInfo)
+    {
+        return str_replace('foo', 'bar', (string)$value);
+    }
+}
+```
+
+### Registering Directives
+
+You can register your directive by appending its class name to the `directives` array on the [registerGqlDirectives](craft3:craft\services\Gql::EVENT_REGISTER_GQL_DIRECTIVES) event object:
+
+```php
+use mynamespace\gql\directives\BarTheFoo;
+use craft\events\RegisterGqlDirectivesEvent;
+use craft\services\Gql;
+use yii\base\Event;
+
+Event::on(
+    Gql::class,
+    Gql::EVENT_REGISTER_GQL_DIRECTIVES,
+    function(RegisterGqlDirectivesEvent $event) {
+        $event->directives[] = BarTheFoo::class;
+    }
+);
+```
+
+## Mutations
+
+A Mutation class defines named mutations that should be available, including consideration for scope access. Each mutation can have its own [arguments](#arguments) and type, and it will need a mutation resolver for modifying data using Craft’s API.
+
+Our example could provide a `saveWidget` mutation for saving a new widget:
+
+::: code
+```graphql{2} GraphQL Query
+mutation NewWidget($title: String) {
+  createWidget(title: $title) {
+    id
+    title
+  }
+}
+
+# query variables:
+# {
+#  "title": "My Glorious Mutated Widget",
+# }
+```
+```json JSON Response
+{
+  "data": {
+    "createWidget": {
+        "id": "123",
+        "title": "My Glorious Mutated Widget"
+    }
+}
+```
+:::
+
+::: tip
+Like the name-getting methods mentioned earlier, Craft’s base elements include a <craft3:craft\base\Element::gqlMutationNameByContext()> method for providing context-sensitive type names.
+:::
+
+### Input Types
```

Comment:
Maybe this needs to be outside of mutations, as input types can be used in queries, too. For example, `relatedTo*` arguments are all input types, too.

Reply:
Moved to “Types” section because that makes way more sense.